presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/PredicatePushDown.java

```diff
@@ -334,7 +334,13 @@ public PlanNode visitUnion(UnionNode node, RewriteContext<Expression> context)
         @Override
         public PlanNode visitFilter(FilterNode node, RewriteContext<Expression> context)
         {
-            return context.rewrite(node.getSource(), combineConjuncts(node.getPredicate(), context.get()));
+            PlanNode rewrittenPlan = context.rewrite(node.getSource(), combineConjuncts(node.getPredicate(), context.get()));
+            if (!(rewrittenPlan instanceof FilterNode)
+                    || !expressionEquivalence.areExpressionsEquivalent(session, ((FilterNode) rewrittenPlan).getPredicate(), node.getPredicate(), types)) {
```

Comment:
maybe it would be more readable of you had `isEqualFilter(PlanNode node)` method which would do all the checks  but not on a single line?